src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,686 @@
+            return;
+        }
+
+        uint256 liquidityBalance = IERC20(liquidityAsset).balanceOf(address(this));
+
+        // Of the claimable withdrawal requests, how much is unclaimed?
+        // That is, the amount of the liquidity assets that is currently allocated for the withdrawal queue
+        uint256 allocatedLiquidity = withdrawsClaimableMem - withdrawsClaimed;
+
+        // If there is no unallocated liquidity assets then there is nothing to add to the queue
+        if (liquidityBalance <= allocatedLiquidity) {
+            return;
+        }
+
+        uint256 unallocatedLiquidity = liquidityBalance - allocatedLiquidity;
+
+        // the new claimable amount is the smaller of the queue shortfall or unallocated weth
+        uint256 addedClaimable = queueShortfall < unallocatedLiquidity ? queueShortfall : unallocatedLiquidity;
+
+        // Store the new claimable amount back to storage
+        withdrawsClaimable = SafeCast.toUint128(withdrawsClaimableMem + addedClaimable);
+    }
+
+    /// @dev Calculate how much of the liquidity asset in the ARM is not reserved for the withdrawal queue.
+    // That is, it is available to be swapped.
+    function _liquidityAvailable() internal view returns (uint256) {
+        // The amount of WETH that is still to be claimed in the withdrawal queue
+        uint256 outstandingWithdrawals = withdrawsQueued - withdrawsClaimed;
+
+        // The amount of the liquidity asset is in the ARM
+        uint256 liquidityBalance = IERC20(liquidityAsset).balanceOf(address(this));
+
+        // If there is not enough liquidity assets in the ARM to cover the outstanding withdrawals
+        if (liquidityBalance <= outstandingWithdrawals) {
+            return 0;
+        }
+
+        return liquidityBalance - outstandingWithdrawals;
+    }
+
+    /// @notice The total amount of assets in the ARM and external withdrawal queue,
+    /// less the liquidity assets reserved for the ARM's withdrawal queue and accrued fees.
+    function totalAssets() public view virtual returns (uint256) {
+        uint256 totalAssetsBeforeFees = _rawTotalAssets();
+
+        // If the total assets have decreased, then we don't charge a performance fee
+        if (totalAssetsBeforeFees <= lastTotalAssets) return totalAssetsBeforeFees;
+
+        // Calculate the increase in assets since the last time fees were calculated
+        uint256 assetIncrease = totalAssetsBeforeFees - lastTotalAssets;
+
+        // Calculate the performance fee and remove from the total assets before new fees are removed
+        return totalAssetsBeforeFees - ((assetIncrease * fee) / FEE_SCALE);
+    }
+
+    /// @dev Calculate the total assets in the ARM, external withdrawal queue,
+    /// less liquidity assets reserved for the ARM's withdrawal queue and past accrued fees.
+    /// The accrued fees are from the last time fees were calculated.
+    function _rawTotalAssets() internal view returns (uint256) {
```

Comment:
I think a different name than rawTotalAssets would be better here, since this isn't the raw total assets. We start with the raw holdings, but then subtract the queue and subtract the fees.

Reply:
yeah, I was never happy with `raw` but it was the best I had at the time.

`_rawTotalAssets` calculates the total assets before accruing any new performance fees. It removes any performance fees that were accrued up to the last time the fees were accrued. This is mostly from the LP functions but can include changing the performance fee rate or fee collector.

`totalAssets` is after any new performance fees are accrued. But `totalAssets` does not persist the new accrued fees. It just dynamically calculates any new performance fees each time.

ChatGPT suggested these new names
* preAccruedTotalAssets
* unaccruedTotalAssets
* grossTotalAssets
* totalAssetsBeforeFees
* baselineTotalAssets
* preFeeTotalAssets
* initialTotalAssets

None of those grab me as `_rawTotalAssets` does include the fees from the last accrual. It also includes the latest assets so its different to `lastTotalAssets`.

I'm leaning towards `preAccruedTotalAssets` but will see if I can think of something better.

Reply:
Hmm, sometimes `_rawTotalAssets` does include the new performance fee. For example, `deposit` calls `_rawTotalAssets` after the performance fee has been accrued and new assets deposited so it can store the value in `lastTotalAssets`.

Reply:
I'm now leaning towards `_availableAssets`

Reply:
I've gone with `_availableAssets()` and also renamed `lastTotalAssets` to `lastAvailableAssets`

Reply:
Seems much better.